### ISSUE TITLE
Projects with same name lead to unintended conflict resolution

### ISSUE THREAD

**Issue by oehme** [gradle/gradle#847]
```
include 'a:foo'
include 'b:foo'
```

```
allprojects {
  apply plugin: 'java'
  group 'org.test'
}

project(':a:foo') {
  dependencies {
    compile project(':b:foo')
  }
}
```
### Expected Behavior
`:a:foo` depends on `:b:foo`. 

### Current Behavior

`:a:foo` depends on itself.

Changing the artifactId using the maven-publish plugin does not help. 

### Context
This effectively prevents users from using short names for projects like `api` or `jdk6`. This will become less of a problem once we support multi-variant projects, but currently it is a common idiom to have separate projects for these concerns.

### Steps to Reproduce (for bugs)
Run `gradle :a:foo:dependencies` to see the problem.

Build scan: https://scans.gradle.com/s/mtomjihobdkfe/dependencies

**Comment by bigdaz** [comment]
To be a bit pedantic, there's not actually any _dependency substitution_ going on here. The problem is that both 'foo' projects produce modules with exactly the same module identifier -`org.test:foo:unspecified`. We use module identifiers to identify nodes in the dependency graph. So the choosing of `a:foo` over `b:foo` is closer to _conflict resolution_ than _dependency substitution_.

Despite this clarification, I fully agree that the current behaviour is unexpected, and a bug.

**Comment by oehme** [comment]
Thanks for the clarification @bigdaz , I changed the title accordingly.

**Comment by markelliot** [comment]
Came across this and was looking for a workaround...so in case anyone else is trying to find a way out: it seems sufficient to change `group` for the subprojects such that they end up with different module coordinates.

**Comment by k-brooks** [comment]
You can also work around this issue by 'aliasing' the problem projects with unique names. Not ideal, but an alternative to changing the group.
`
include ':a:foo-a', ':b:foo-b'
project( ':a:foo-a').projectDir = file('a/foo')
project( ':b:foo-b').projectDir = file('b/foo')
`

**Comment by milosmns** [comment]
There is one more workaround I came across (also done in `settings.gradle` after including the modules using their directory names):
```groovy
findProject(":moduleDirectory")?.name = 'your-new-module-name'
```
And then you can reference it in Gradle files using `your-new-module-name`.

**Comment by aaronabf** [comment]
+1, this is very detrimental to our uses

**Comment by christofvanhove** [comment]
+1

**Comment by blrichards** [comment]
+1

**Comment by UkonnRa** [comment]
still exists...

**Comment by theaspect** [comment]
@melix will this issue be backported to 5.x branch?

**Comment by melix** [comment]
No, there's no such plan, we currently don't maintain older branches.

**Comment by ljacomet** [comment]
The fix made for Gradle 6.2 has too many unintended consequences on builds out there. So it will be reverted in 6.2.1 and a different fix will have to be designed.

**Comment by ivan-osipov** [comment]
@ljacomet what is the best workaround for now from your perspective?

**Comment by ljacomet** [comment]
The best thing you can do for now is make sure all projects have different `group:name` in your multi project, by setting the `group` at the project level and not only in the publication.

If you can play with the `group` part to disambiguate, do that, it will allow you to keep simple `name`.
If that's not an option, you will have to disambiguate at the `name` level, possibly making invoking tasks on some projects longer.

For example:
```
myApp
  - client
    - common
  - server
    - common
```

If you configure all projects with the group `com.app`, you will have conflicts between the two `common` subprojects.
Make sure you have either
* `com.app.client:common` and `com.app.server:common`, through setting `group` in each project 
* or `com.app:client-common` and `com.app:server-common`, through explicit naming in `settings.gradle`.

**Comment by LesTR** [comment]
Hi, we are getting into same issue right now also with latest gradle (7.4.1). Is there any plan to fix it?

**Comment by pedrolamarao** [comment]
Just been hit by this. Very unexpected.

**Comment by gmazzo** [comment]
Any plan to address this soon? I has been reported for years now, and it seems to be serious enough for a mature product 🤔 

I didn't check the code, but it seems to be related to the default project capability, right? If this is just for internal resolution, couldn't a quick fix the a to use the project's `path` instead of its `name`?

**Comment by Jire** [comment]
This should absolutely be addressed, it's been an issue for years.

Here is my workaround for your `rootProject`'s `build.gradle.kts` that sets group:

```kotlin
group = "com.example"

subprojects {
    val subGroup = path.replace(':', '.').replace('-', '_')
    group = "${rootProject.group}$subGroup"
}
```

**Comment by gmazzo** [comment]
I got hit by this again:
<img width="278" alt="image" src="https://user-images.githubusercontent.com/513566/181773561-58ebb2da-dd8f-40a3-a791-0ff79e1ec2fb.png">
It was making me crazy when I realized by chance that the scan revealed the "confusion".

It has been 6 years since this was reported and Gradle still "confuses" different modules with the same local name when doing resolution of local modules dependencies. 

Come on, you can do better. Please prioritize this ticket.

**Comment by rafallezanko** [comment]
Hi, 6 years passed... Do you have plan to solve it?

**Comment by melix** [comment]
I don't know if there are "plans to solve it", but I can explain why it works like this. Basically, Gradle supports conflict resolution between a project and its published binary equivalent. For example, you can have a project which depends on a different, published, version of itself. There are users who depend on this behavior, which is also tweakable (for example, to always prefer the project dependency _or_ to always prefer the binary dependency). 

In order to support this, and also to support composite builds which replace binary dependencies with project dependencies, Gradle needs to map a project to its group, artifact, version (GAV) coordinates. If you do nothing, like in the example above, both projects, despite being different, will end up having the same GAV coordinates. Therefore, during graph resolution, they are considered the _same_.

The workaround is therefore to make sure that you have different coordinates for both projects, by changing the group id as mentioned in https://github.com/gradle/gradle/issues/847#issuecomment-1159365667.

Solving this effectively requires a better mapping model between projects and GAV coordinates, for example something which automatically deduces GAV coordinates by convention.

**Comment by eygraber** [comment]
I think something in Kotlin 1.8 or Gradle 8.0 RC is making this worse because I have projects that worked previously, but are failing with these errors after updating.

Changing the `group` does work for me.

**Comment by bcmedeiros** [comment]
> I recently stumbled into this issue while refactoring a larger project and I am somewhat reluctant to give up the benefits of a tree-based structure and short names for subprojects.

I'm on the same boat here, I have a repo with many modules in a nested hierarchy, so similar modules will have the same sub-modules names and it's very annoying having to use a custom name for all the dependencies of those sub-modules.

I just ran the following code in a small repro project here:

```
include(":a:lib")
val aLibProject = project(path = ":a:lib") // Path so far is ":a:lib"
aLibProject.name = "a-lib" // and then I change the name
println(aLibProject.path) // This prints ":a:a-lib" instead of the original ":a:lib", which makes no sense at all.
project(path = ":a:lib") // FAIL, path is no longer ":a:lib" (???)
```

This seems to be the actual intended behavior based on this code:
https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/initialization/DefaultProjectDescriptor.java#L108

Does someone now why is that? If I include a project with `include(":a:lib")`, the path should be `":a:lib"` no matter what, I don't understand why Gradle would change the path if I change the name. The default name is the last component of the path, I get that, but if we change the name, the name should be the new one while the path should be the original one.

**Comment by NikolayMetchev** [comment]
Should this issue be closed? It seems to have started working at some point. Gradle 7.6 works for me.

**Comment by bcmedeiros** [comment]
Issue is 100% reproducible for me, just created a new repo with it: https://github.com/bcmedeiros/gradle-issue-847

```
FAILURE: Build failed with an exception.

* What went wrong:
Circular dependency between the following tasks:
:b:lib:classes
\--- :b:lib:compileJava
     +--- :b:lib:compileKotlin
     |    \--- :b:lib:jar
     |         +--- :b:lib:classes (*)
     |         +--- :b:lib:compileJava (*)
     |         \--- :b:lib:compileKotlin (*)
     \--- :b:lib:jar (*)

(*) - details omitted (listed previously)
```

**Comment by sschuberth** [comment]
> Here is my workaround for your `rootProject`'s `build.gradle.kts` that sets group for all subprojects automatically:

Unfortunately, setting just the group name is not enough if you use the [distribution plugin](https://docs.gradle.org/current/userguide/distribution_plugin.html) and have nested directories with project directories that are named the same and thus lead to identical default project names, and thus conflicting JAR names which cannot be put into the same flat `lib` directory...

**Comment by nbrugger-tgm** [comment]
Do i need to downgrade to Gradle 7 now :( ? Are there any other workarounds, setting a unique `group` doesn't work for me.

**Comment by portlek** [comment]
i've just made something like this:
```kotlin
project(":services").children.forEach { m -> m.children.forEach { s -> s.name = "${m.name}-${s.name}" } }
```
every project is different so you should try to design your maven packages/GAV first. for instance:
This settings.gradle.kts change looks like this in my maven local repository:
![image](https://user-images.githubusercontent.com/12835420/234217591-014e9fb4-a0d7-43b2-b547-242b7f27b07d.png)
and i like the way it looks like so, i kept like that.
The only downside is the one unnecessary thing when you use it via project() in dependencies:
```kotlin
//                             .......
api(project(":services:profile:profile-common"))
```
but i think this doesn't really matter if you fix the actual issue that does not event let you build the project, so.

**Comment by AndrewShinCoder** [comment]
did you try type.null/project(":services").children.forEach { m -> m.children.forEach { s -> s.name = "${m.name}-${s.name}" } }[name/server/run{beta}]
{
         }

<CC.project(":services").children.forEach { m -> m.children.forEach { s -> s.name = "${m.name}-${s.name}" } }<error.c=denyFAILURE: Build failed with an exception.

* What went wrong:
Circular dependency between the following tasks:
:b:lib:classes
\--- :b:lib:compileJava
     +--- :b:lib:compileKotlin
     |    \--- :b:lib:jar
     |         +--- :b:lib:classes (*)
     |         +--- :b:lib:compileJava (*)
     |         \--- :b:lib:compileKotlin (*)
     \--- :b:lib:jar (*)

(*) - details omitted (listed previously)>

<end/s>
{
     }
          } 
               }

**Comment by SakurajimaMaii** [comment]
+1 This affected me too，I hope **log:core** depends on **core** but I encounter the problem shown in the figure
![image](https://github.com/user-attachments/assets/289522b2-8892-4eaa-af2f-fcc0c21a13da)

**Comment by SgtSilvio** [comment]
Stumbled upon this issue while figuring out again what is the best practice for naming Gradle subprojects.
My current conclusion is that you want
- "fully qualified"/"absolute" project names (not just `api` or `core`, but `my-module-api` and `my-module-core`)
- short names in project paths (not something like `:my-module:my-module-api`, but `:my-module:api`).

Because both above points can currently not be easily configured with Gradle, I created a small settings plugin that makes tries to achieve something as close as possible: https://github.com/SgtSilvio/gradle-structure.

### Explanation why I came to this conclusion
A project name should be absolute, while a project path is by definition hierarchical (even if you only have one layer of subprojects, they are still hierarchically grouped under the root project). Thus, the project path should not just be a concatenation of the project names.
Moreover, the project name and path serve a different purpose. The project name is conventionally used for module coordinates, artifact names, capabilities, and by many third-party plugins (for example by the Kotlin plugin for the module name), so it affects the external identity of a project. A project path is solely for internal use in your Gradle build, like defining project dependencies or selecting task paths.

### Suggestion for a solution
Currently, the project path basically defines the project group and name, but only the group can currently be set to a value that is independent of the project path. This is also what is causing this issue. Being forced to set different groups for projects that have the same last path segment is not a solution, only a workaround the actual issue that you want to set the project name independent of the last segment in the project path.

For the following consider this example settings file:
```kotlin
rootProject.name = "example-app"
include("server")
include("server:api")
```

This results in the following project paths, groups and names:
| project path | project group | project name |
|---|---|---|
| `:` |  | `example-app` |
| `:server` | `example-app` | `server` |
| `:server:api` | `example-app.server` | `api` |

Let's say, that what what you actually want is this:
| project path | project group | project name |
|---|---|---|
| `:` | `org.example` | `example-app` |
| `:server` | `org.example` | `example-app-server` |
| `:server:api` | `org.example` | `example-app-server-api` |

So you set the group in every project to `org.example` and you get this:
| project path | project group | project name |
|---|---|---|
| `:` | `org.example` | `example-app` |
| `:server` | `org.example` | `server` |
| `:server:api` | `org.example` | `api` |

Now, you might realize that you can not set the project name without changing the path. Or you don't realize the issue of the unqualified project names and you might run into this issue here.

My suggestion for a solution would be to allow to set the project name and the "project path segment name" separately. This could also be done in a way that is fully backwards compatible, by setting both the project name and "project path segment name" to the same value by default. In a major Gradle release, this default behavior could then be changed to a convention that is better for the most use cases.

**Comment by SgtSilvio** [comment]
A remark regarding the fix that was reverted
> The fix made for Gradle 6.2 has too many unintended consequences on builds out there. So it will be reverted in 6.2.1 and a different fix will have to be designed.

My suggested solution removes the need for any special handling of project dependencies. Imho, using "module identifiers to identify nodes in the dependency graph" is the right thing to do. Also, using the project group and name as the module coordinates is a sane convention. Decoupling the project's segment name in the path should be a much less intrusive change.

**Comment by IAmShaarawy** [comment]
i wasted a day and half of my time because of this issue and it is there for almost 9 years.

**Comment by bcmedeiros** [comment]
> i wasted a day and half of my time because of this issue and it is there for almost 9 years.

That's what happens when something is badly designed. I honestly gave up on creating any modules with short names, I just repeat the common prefix:

a
a-lib-contract
a-lib-impl
a-app
b
b-lib-contract
b-lib-impl
b-app

I honestly moved on from this, anything that is not a flat list of folders in Gradle is just an illusion, I wish they could just acknowledge and document that instead of pretending it works or that this issue will be ever fixed and making people waste their time.

**Comment by bcmedeiros** [comment]
> My suggestion for a solution would be to allow to set the project name and the "project path segment name" separately. 

@SgtSilvio you reached the same conclusion as I did when I delved into this 2 years ago and raised https://github.com/gradle/gradle/issues/23848

The issue was closed by @eskatos with a "this is the way it works" justification 🤔
https://github.com/gradle/gradle/issues/23848#issuecomment-1439749728

Yeah, I know this is the way it works, and it's obviously broken, but I guess he didn't even care checking the related issue which is this one.

**Comment by SgtSilvio** [comment]
@bcmedeiros thanks for replying to my comment. Already thought that nobody would read it though it took me some time to write.
I can understand the frustration when your issues/PRs get closed as "won't do". In my experience this is almost always the case because of some misunderstanding + time investment constraints.
I guess, that in your case, your PR/issue got closed because you suggested to "not affect project path when name is changed", which would basically change an API contract. Then I understand the argument "this is the way how the API works".
My suggestion for a solution that hopefully the Gradle engineers can agree with, would be to not change any API (so `project(":server").name = "example-app-server"` would not change in behavior), but introduce a way that allows us to set the "project path segment name". This would be a new API and without using it, there would be no change in behavior.

For the example from above this could look like the following:
```
rootProject.name = "example-app"

include("server")
project(":server").apply {
    name = "example-app-server"
    pathSegmentName = "server"
}
include("server:api")
project(":server:api").apply {
    name = "example-app-server-api
    pathSegmentName = "api"
}
```
(I don't suggest a final name for `pathSegmentName` yet, I just used this one for now to showcase what I mean.)

Adding the `pathSegmentName` property would be the only change that is sufficient to solve this issue, no other APIs would change. We could also add other APIs that make this more convenient (like introducing a new include method that already configures the above settings as a convention), but they are not required and would only sit on top, so let's please focus on the `pathSegmentName` property (or whatever we call it, `pathName`, `nameInPath`, ...).
@eskatos as you have already been mentioned, what do you think about the above suggestion?

**Comment by bcmedeiros** [comment]
@SgtSilvio yeah, I was not attached to my PR at all, but after my issue was closed without any discussions I concluded it would be pointless to keep discussing it. My issue was closed, not my PR.

**Comment by StormWhisper** [comment]
Still a problem in Gradle 8.12.1. Hope it will be fixed someday

**Comment by illarionov** [comment]
> Could Gradle detect ambiguously named projects and log an informative message to help users immediately understand the problem? Presently there's no useful feedback, it just mysteriously doesn't work.

Just wanted to note that there’s also issue #16986 for this kind of warning.

**Comment by fellahst** [comment]
I filed a ticket related about this bug in Kotlin YouTrack https://youtrack.jetbrains.com/issue/KT-82662/Kotlin-Gradle-Plugin-archivesTaskOutputAsFriendModule-property-ignored-when-multiple-modules-share-same-folder-name-in-different .
